### PR TITLE
CB-11747 External database create flow is not restartable

### PR DESCRIPTION
**Pull request by bergerdenes**
In case flow was stopped and then restarted just after the external
DB provision started, it might happen that upon launching a new
create request to RedBeams caused and exception. This is fixed now.

See detailed description in the commit message.

### CODE REVIEW

File: redbeams-api/src/main/java/com/sequenceiq/redbeams/api/endpoint/v4/databaseserver/DatabaseServerV4Endpoint.java
Comment (lajosrodek) [inline redbeams-api/src/main/java/com/sequenceiq/redbeams/api/endpoint/v4/databaseserver/DatabaseServerV4Endpoint.java:89]: Note: I think this will have to be updated soon to permit any non-environment CRNs for sake of looking up DB servers for non-CB "clusters" (e.g. DFX). See [this](https://github.com/hortonworks/cloudbreak/pull/10459/files#diff-11d87b3ace4084117083bb0317790b6c0f709d00603e662c2a6d9581247ce5c0) change in #10459.

CC @gergopapi2

Comment (bergerdenes) [reply redbeams-api/src/main/java/com/sequenceiq/redbeams/api/endpoint/v4/databaseserver/DatabaseServerV4Endpoint.java:89]: but does it block the merge?

Comment (lajosrodek) [reply redbeams-api/src/main/java/com/sequenceiq/redbeams/api/endpoint/v4/databaseserver/DatabaseServerV4Endpoint.java:89]: Certainly not, I just flagged it as something to be fixed once both PRs got merged. 😄

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/externaldatabase/ExternalDatabaseService.java
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/externaldatabase/ExternalDatabaseService.java:137]: drop this line

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/externaldatabase/ExternalDatabaseService.java
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/externaldatabase/ExternalDatabaseService.java:142]: add return here

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/externaldatabase/ExternalDatabaseService.java
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/externaldatabase/ExternalDatabaseService.java:144]: add debug or info log here and return with `Optional#empty`

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/externaldatabase/ExternalDatabaseService.java
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/externaldatabase/ExternalDatabaseService.java:78]: request creation should be moved right before this line

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/externaldatabase/ExternalDatabaseService.java
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/externaldatabase/ExternalDatabaseService.java:69]: this should be inside the try. also would be nicer if the `if` part would be in a separate method and there would be only one value assignment for this variable

File: core/src/main/java/com/sequenceiq/cloudbreak/service/rdsconfig/RedbeamsClientService.java
Comment (lacikaaa) [inline core/src/main/java/com/sequenceiq/cloudbreak/service/rdsconfig/RedbeamsClientService.java:89]: move to a validate method please

File: redbeams-api/src/main/java/com/sequenceiq/redbeams/api/endpoint/v4/databaseserver/DatabaseServerV4Endpoint.java
Comment (lacikaaa) [reply redbeams-api/src/main/java/com/sequenceiq/redbeams/api/endpoint/v4/databaseserver/DatabaseServerV4Endpoint.java:89]: as the API can't be modified later I would say it would be better to call this resourceCrn or similar which can be meaningful for other services. But I will leave it to @gergopapi2

File: redbeams/src/main/java/com/sequenceiq/redbeams/service/dbserverconfig/DatabaseServerConfigService.java
Comment (lacikaaa) [inline redbeams/src/main/java/com/sequenceiq/redbeams/service/dbserverconfig/DatabaseServerConfigService.java:225]: you could use `Optional#orElseThrow` to get the value and throw exception

**Review by lacikaaa — CHANGES_REQUESTED** [review]
please add unit test for the changes in `ExternalDatabaseService`

File: redbeams-api/src/main/java/com/sequenceiq/redbeams/api/endpoint/v4/databaseserver/DatabaseServerV4Endpoint.java
Comment (doktoric) [reply redbeams-api/src/main/java/com/sequenceiq/redbeams/api/endpoint/v4/databaseserver/DatabaseServerV4Endpoint.java:89]: resourceCrn would be nicer yes because other services also want to use redbeams

Comment (gergopapi2) [reply redbeams-api/src/main/java/com/sequenceiq/redbeams/api/endpoint/v4/databaseserver/DatabaseServerV4Endpoint.java:89]: I do agree, however the name clusterCrn is already out in the AllocateDatabaseServerV4Request. For consistency it would make sense to keep the name I think.

Comment (bergerdenes) [reply redbeams-api/src/main/java/com/sequenceiq/redbeams/api/endpoint/v4/databaseserver/DatabaseServerV4Endpoint.java:89]: leaved as `clusterCrn` but changed the `@ValidCrn` as per @lajosrodek  suggested

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/externaldatabase/ExternalDatabaseService.java
Comment (bergerdenes) [reply core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/externaldatabase/ExternalDatabaseService.java:137]: if possible I like the single point of return, but, whatever. 🤷🏼‍♂️

File: redbeams/src/main/java/com/sequenceiq/redbeams/service/dbserverconfig/DatabaseServerConfigService.java
Comment (bergerdenes) [reply redbeams/src/main/java/com/sequenceiq/redbeams/service/dbserverconfig/DatabaseServerConfigService.java:225]: nit, but ok

File: core/src/main/java/com/sequenceiq/cloudbreak/service/rdsconfig/RedbeamsClientService.java
Comment (bergerdenes) [reply core/src/main/java/com/sequenceiq/cloudbreak/service/rdsconfig/RedbeamsClientService.java:89]: ok

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/externaldatabase/ExternalDatabaseService.java
Comment (bergerdenes) [reply core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/externaldatabase/ExternalDatabaseService.java:69]: not possible to move the variable declaration into the `try` block as the `waitAndGetDatabase` needs it too and that is outside the try `block`

File: core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/externaldatabase/ExternalDatabaseService.java
Comment (bergerdenes) [reply core/src/main/java/com/sequenceiq/cloudbreak/core/flow2/externaldatabase/ExternalDatabaseService.java:78]: ok

**Comment by bergerdenes** [comment]
unit test added

File: redbeams/src/main/java/com/sequenceiq/redbeams/service/dbserverconfig/DatabaseServerConfigService.java
Comment (lacikaaa) [inline redbeams/src/main/java/com/sequenceiq/redbeams/service/dbserverconfig/DatabaseServerConfigService.java:231]: it would be nice to follow the repository's pattern here: `find` returns `Optional`, while `get` returns the pojo and fails in case there is none. so the method name would look like here: `findByEnvironmentIdAndClusterCrn`, same as the repo

**Review by lacikaaa — APPROVED** [review]
that only naming nit is left, otherwise LGTM, feel free to merge after fixing that one. thanks for the changes

File: redbeams/src/main/java/com/sequenceiq/redbeams/service/dbserverconfig/DatabaseServerConfigService.java
Comment (doktoric) [reply redbeams/src/main/java/com/sequenceiq/redbeams/service/dbserverconfig/DatabaseServerConfigService.java:231]: findByEnvironmentCrnAndClusterCrn right ?

Comment (lacikaaa) [reply redbeams/src/main/java/com/sequenceiq/redbeams/service/dbserverconfig/DatabaseServerConfigService.java:231]: right, I copied the repository method name. unfortunately in redbeams it called `environmentId` instead of crn...

Comment (doktoric) [reply redbeams/src/main/java/com/sequenceiq/redbeams/service/dbserverconfig/DatabaseServerConfigService.java:231]: I see you @lacikaaa . I see you ...

**Pull request merged**
